### PR TITLE
Consider max priority in CustomWeighting

### PR DESCRIPTION
**Pull request by karussell**
See #2334.

Similar to maxSpeed we need to consider priority for getMinWeight. Especially when we introduce #2333 that uses an EncodedValue with a range from 0 to 1.5

### CODE REVIEW

File: core/src/main/java/com/graphhopper/routing/weighting/custom/CustomModelParser.java [hunk 1]
```diff
@@ -409,45 +411,44 @@ private static SimpleCompiler createCompiler(long counter, Java.AbstractCompilat
...
-    static double findMaxSpeed(List<Statement> speedStatements, final double maxSpeed) {
+    static double findMaxSpeed(List<Statement> statements, final double maxSpeed) {
         // throw an error if one of the limit statements won't possibly do anything
-        Optional<Statement> falseStatement = speedStatements.stream()
+        Optional<Statement> falseStatement = statements.stream()
                 .filter(st -> LIMIT.equals(st.getOperation()) && st.getValue() > maxSpeed)
                 .findFirst();
         if (falseStatement.isPresent())
             throw new IllegalArgumentException("Can never apply 'limit_to': " + falseStatement.get().getValue()
                     + " because maximum vehicle speed is " + maxSpeed);
```
[the remark this answers is not in the document]
Comment (karussell) [reply core/src/main/java/com/graphhopper/routing/weighting/custom/CustomModelParser.java:430]: Why do you think would we need a maximum priority?

File: core/src/main/java/com/graphhopper/routing/weighting/custom/CustomModelParser.java [hunk 2]
```diff
@@ -113,7 +113,9 @@ public static CustomWeighting createWeighting(FlagEncoder baseFlagEncoder, Encod
             // The class does not need to be thread-safe as we create an instance per request
             CustomWeightingHelper prio = (CustomWeightingHelper) clazz.getDeclaredConstructor().newInstance();
             prio.init(lookup, avgSpeedEnc, customModel.getAreas());
-            return new CustomWeighting.Parameters(prio::getSpeed, prio::getPriority, findMaxSpeed(customModel, globalMaxSpeed),
+            return new CustomWeighting.Parameters(prio::getSpeed, prio::getPriority,
+                    findMaxSpeed(customModel.getSpeed(), globalMaxSpeed),
+                    findMaxWithoutCheck(customModel.getPriority(), 1),
```
[the remark this answers is not in the document]
Comment (karussell) [reply core/src/main/java/com/graphhopper/routing/weighting/custom/CustomModelParser.java:118]: Yes, I think I will introduce getEncodedValue(<encoder$priority>).getMaxDecimal() in #2333 when this PR here is merged (?)

File: core/src/main/java/com/graphhopper/routing/weighting/custom/CustomModelParser.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (easbar) [reply core/src/main/java/com/graphhopper/routing/weighting/custom/CustomModelParser.java:430]: No this checks simply checks that we do not do something like `speed: limit_to: 250` when max speed is only `140` anyway.

File: core/src/main/java/com/graphhopper/routing/weighting/custom/CustomModelParser.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (easbar) [reply core/src/main/java/com/graphhopper/routing/weighting/custom/CustomModelParser.java:118]: Hm for now I would just do `encoder.getMaxPriority`, its almost exactly the same as `encoder.getMaxSpeed`?

[the remark this answers is not in the document]
Comment (easbar) [reply core/src/main/java/com/graphhopper/routing/weighting/custom/CustomModelParser.java:118]: But attaching the 'max value' information seems useful either way :+1:

File: core/src/main/java/com/graphhopper/routing/weighting/custom/CustomModelParser.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (karussell) [reply core/src/main/java/com/graphhopper/routing/weighting/custom/CustomModelParser.java:430]: The maximum speed is something physical that should not exceed practical values as ETA is affected, whereas the priority is something arbitrary / subjective - i.e. it can be 5 million without a problem. The limit of 1.5 is only in our EncodedValue (`encoder$priority`) which has not much to do with the general maximum priority here except that we will use 1.5 as default maximum if no other is found.

So basically when we start with 1.5 and unconditionally increase it with a factor of 10 and then do another unconditional `limit_to:5` should not be a problem (although illogical to have both in the same custom model)

File: core/src/main/java/com/graphhopper/routing/weighting/custom/CustomModelParser.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (karussell) [reply core/src/main/java/com/graphhopper/routing/weighting/custom/CustomModelParser.java:118]: I would not want to introduce this method to the encoder when we want to resolve the FlagEncoder class and already have the info easily via the EncodedValue directly.

File: core/src/main/java/com/graphhopper/routing/weighting/custom/CustomModelParser.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (karussell) [reply core/src/main/java/com/graphhopper/routing/weighting/custom/CustomModelParser.java:430]: The following use case makes more sense and should work on the server-side:

```
{
  "priority" [{
      "if": "xy",
      "multiply_by": 5
  }, {
      "else": "",
      "multiply_by": 10
  }, {
      "if": "true",
      "limit_to": 5
  }
}
```

But it won't work if we use the "max storage" value of `encoder$priority` as the maximum possible value for the custom model priority.

[the remark this answers is not in the document]
Comment (easbar) [reply core/src/main/java/com/graphhopper/routing/weighting/custom/CustomModelParser.java:436]: Ok, right. The check does not stop you from using `multiply_by: 5_000_000`, but you if we enable you can no longer first multiply with a large number and then limit to something above the storage limit.

But the thing I do not understand is why we should be allowed to do this for priority and not for speed. For example what if we want an electric bike (assuming storage max speed 30) like this:

```json
{
  "speed": [
      {
          "if": "surface != cobblestone",
          "multiply_by": 1.5
      },
      {
          "if": "true",
          "limit_to": 35
      }
  ]
}
```

File: core/src/main/java/com/graphhopper/routing/weighting/custom/CustomModelParser.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (easbar) [reply core/src/main/java/com/graphhopper/routing/weighting/custom/CustomModelParser.java]: Ok.

File: core/src/main/java/com/graphhopper/routing/weighting/custom/CustomModelParser.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (karussell) [reply core/src/main/java/com/graphhopper/routing/weighting/custom/CustomModelParser.java:436]: Thanks. This made me think :)
Currently the check would let you know that the limit_to will never have an effect because we use the storage max speed to reduce at the end of every speed calculation (probably I thought this is required because of LM). So we should just replace this storage-maxSpeed and use the result of findMaxSpeed if any. There are multiple advantages: we can increase the max speed via server-side config (nice!) and it would be better backward compatible (because every max speed reduction in storage could result in an exception for previously working custom model queries). 

Will change!

[the remark this answers is not in the document]
Comment (karussell) [reply core/src/main/java/com/graphhopper/routing/weighting/custom/CustomModelParser.java]: Can you have a look into the last commit?

File: core/src/main/java/com/graphhopper/routing/DefaultWeightingFactory.java [hunk 3]
```diff
@@ -75,8 +75,9 @@ public Weighting createWeighting(Profile profile, PMap requestHints, boolean dis
                 throw new IllegalArgumentException("custom weighting requires a CustomProfile but was profile=" + profile.getName());
             CustomModel queryCustomModel = requestHints.getObject(CustomModel.KEY, null);
             CustomProfile customProfile = (CustomProfile) profile;
-            queryCustomModel = queryCustomModel == null ?
-                    customProfile.getCustomModel() : CustomModel.merge(customProfile.getCustomModel(), queryCustomModel);
+            if (queryCustomModel != null)
+                queryCustomModel.check(customProfile.getCustomModel(), encoder.getMaxSpeed());
```
Comment (easbar) [inline core/src/main/java/com/graphhopper/routing/DefaultWeightingFactory.java:79]: What if `customProfile.getCustomModel().findMaxSpeed(encoder.getMaxSpeed()) < encoder.getMaxSpeed())`? Right now I'm not sure if we should use the server-side custom model's max speed as input to `queryCustomModel.check(_, maxSpeed)` instead?

Comment (karussell) [reply core/src/main/java/com/graphhopper/routing/DefaultWeightingFactory.java:79]: Oh, yes. This is correct. All a bit interweaved :)

File: web/src/test/java/com/graphhopper/http/resources/RouteResourceCustomModelTest.java [hunk 4]
```diff
@@ -242,34 +230,71 @@ public void testSubnetworkRemovalPerProfile() {
...
+                "   \"speed\": [{" +
+                "       \"if\": \"true\", \"limit_to\": 55" +
+                "     }] " +
+                " }," +
+                " \"ch.disable\": true" +
+                "}";
+
+        final Response response = query(jsonQuery, 400);
+        JsonNode json = response.readEntity(JsonNode.class);
+        assertTrue(json.get("message").asText().startsWith("vehicle speed of CustomModel in query cannot be bigger than 40"), json.get("message").asText());
```
Comment (karussell) [inline web/src/test/java/com/graphhopper/http/resources/RouteResourceCustomModelTest.java:268]: With all the info I have now I would rather remove this error and silently ignore statements with a too big limit_to value, to easier maintain backward compatibility. Because whenever we increase the vehicle's max_speed we could break query-side custom models.

WDYT @easbar?

Comment (easbar) [reply web/src/test/java/com/graphhopper/http/resources/RouteResourceCustomModelTest.java:268]: To be honest I never saw so much value in this check. Your point about backward compatibility seems reason enough to remove it with my current understanding.

One issue that this check somewhat helps with is that the max speed is hardly visible for the user, but I am sure there are better ways to improve this than having this check.
